From: Lucas Stach <l.stach@pengutronix.de>
Date: Thu, 15 Sep 2022 15:48:07 +0200
Subject: [PATCH] arm64: dts: imx8mp: describe DMA capable memory

Without a dma-ranges property the kernel tries to infer the DMA addressing
restrictions from internal information available about the devices and
busses and sometimes gets it wrong.

Instead of relying on those fallback heuristics, describe the DMA capable
memory range in the DT. As the size of the DMA address space is 8GB in
total the soc bus needs to be switched to 64bit DT addresses, in order to
be able to describe this dma-range.

Signed-off-by: Lucas Stach <l.stach@pengutronix.de>
---
 arch/arm64/boot/dts/freescale/imx8mp.dtsi | 43 +++++++++++++++++--------------
 1 file changed, 23 insertions(+), 20 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/imx8mp.dtsi b/arch/arm64/boot/dts/freescale/imx8mp.dtsi
index cf0228124da7..edd0d76e769a 100644
--- a/arch/arm64/boot/dts/freescale/imx8mp.dtsi
+++ b/arch/arm64/boot/dts/freescale/imx8mp.dtsi
@@ -296,9 +296,11 @@ timer {
 
 	soc: soc@0 {
 		compatible = "fsl,imx8mp-soc", "simple-bus";
-		#address-cells = <1>;
-		#size-cells = <1>;
-		ranges = <0x0 0x0 0x0 0x3e000000>;
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges = <0x0 0x0 0x0 0x0 0x3e000000 0x0>;
+		dma-ranges = <0x0 0x40000000 0x0 0x40000000 0x2 0x0>;
+
 		nvmem-cells = <&imx8mp_uid>;
 		nvmem-cell-names = "soc_unique_id";
 
@@ -1029,7 +1031,7 @@ eqos: ethernet@30bf0000 {
 
 		noc: interconnect@32700000 {
 			compatible = "fsl,imx8mp-noc", "fsl,imx8m-noc";
-			reg = <0x32700000 0x100000>;
+			reg = <0x0 0x32700000 0x0 0x100000>;
 			clocks = <&clk IMX8MP_CLK_NOC>;
 			#interconnect-cells = <1>;
 			operating-points-v2 = <&noc_opp_table>;
@@ -1292,7 +1294,8 @@ hdmi_tx_phy: phy@32fdff00 {
 
 		pcie: pcie@33800000 {
 			compatible = "fsl,imx8mp-pcie";
-			reg = <0x33800000 0x400000>, <0x1ff00000 0x80000>;
+			reg = <0x0 0x33800000 0x0 0x400000>,
+			      <0x0 0x1ff00000 0x0 0x80000>;
 			reg-names = "dbi", "config";
 			#address-cells = <3>;
 			#size-cells = <2>;
@@ -1323,7 +1326,7 @@ pcie: pcie@33800000 {
 
 		gpu3d: gpu@38000000 {
 			compatible = "vivante,gc";
-			reg = <0x38000000 0x8000>;
+			reg = <0x0 0x38000000 0x0 0x8000>;
 			interrupts = <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&clk IMX8MP_CLK_GPU3D_ROOT>,
 				 <&clk IMX8MP_CLK_GPU3D_SHADER_CORE>,
@@ -1340,7 +1343,7 @@ gpu3d: gpu@38000000 {
 
 		gpu2d: gpu@38008000 {
 			compatible = "vivante,gc";
-			reg = <0x38008000 0x8000>;
+			reg = <0x0 0x38008000 0x0 0x8000>;
 			interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&clk IMX8MP_CLK_GPU2D_ROOT>,
 				 <&clk IMX8MP_CLK_GPU_ROOT>,
@@ -1354,8 +1357,8 @@ gpu2d: gpu@38008000 {
 
 		gic: interrupt-controller@38800000 {
 			compatible = "arm,gic-v3";
-			reg = <0x38800000 0x10000>,
-			      <0x38880000 0xc0000>;
+			reg = <0x0 0x38800000 0x0 0x10000>,
+			      <0x0 0x38880000 0x0 0xc0000>;
 			#interrupt-cells = <3>;
 			interrupt-controller;
 			interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_HIGH>;
@@ -1364,19 +1367,19 @@ gic: interrupt-controller@38800000 {
 
 		edacmc: memory-controller@3d400000 {
 			compatible = "snps,ddrc-3.80a";
-			reg = <0x3d400000 0x400000>;
+			reg = <0x0 0x3d400000 0x0 0x400000>;
 			interrupts = <GIC_SPI 147 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
 		ddr-pmu@3d800000 {
 			compatible = "fsl,imx8mp-ddr-pmu", "fsl,imx8m-ddr-pmu";
-			reg = <0x3d800000 0x400000>;
+			reg = <0x0 0x3d800000 0x0 0x400000>;
 			interrupts = <GIC_SPI 98 IRQ_TYPE_LEVEL_HIGH>;
 		};
 
 		usb3_phy0: usb-phy@381f0040 {
 			compatible = "fsl,imx8mp-usb-phy";
-			reg = <0x381f0040 0x40>;
+			reg = <0x0 0x381f0040 0x0 0x40>;
 			clocks = <&clk IMX8MP_CLK_USB_PHY_ROOT>;
 			clock-names = "phy";
 			assigned-clocks = <&clk IMX8MP_CLK_USB_PHY_REF>;
@@ -1388,8 +1391,8 @@ usb3_phy0: usb-phy@381f0040 {
 
 		usb3_0: usb@32f10100 {
 			compatible = "fsl,imx8mp-dwc3";
-			reg = <0x32f10100 0x8>,
-			      <0x381f0000 0x20>;
+			reg = <0x0 0x32f10100 0x0 0x8>,
+			      <0x0 0x381f0000 0x0 0x20>;
 			clocks = <&clk IMX8MP_CLK_HSIO_ROOT>,
 				 <&clk IMX8MP_CLK_USB_ROOT>;
 			clock-names = "hsio", "suspend";
@@ -1403,7 +1406,7 @@ usb3_0: usb@32f10100 {
 
 			usb_dwc3_0: usb@38100000 {
 				compatible = "snps,dwc3";
-				reg = <0x38100000 0x10000>;
+				reg = <0x0 0x38100000 0x0 0x10000>;
 				clocks = <&clk IMX8MP_CLK_HSIO_AXI>,
 					 <&clk IMX8MP_CLK_USB_CORE_REF>,
 					 <&clk IMX8MP_CLK_USB_ROOT>;
@@ -1418,7 +1421,7 @@ usb_dwc3_0: usb@38100000 {
 
 		usb3_phy1: usb-phy@382f0040 {
 			compatible = "fsl,imx8mp-usb-phy";
-			reg = <0x382f0040 0x40>;
+			reg = <0x0 0x382f0040 0x0 0x40>;
 			clocks = <&clk IMX8MP_CLK_USB_PHY_ROOT>;
 			clock-names = "phy";
 			assigned-clocks = <&clk IMX8MP_CLK_USB_PHY_REF>;
@@ -1430,8 +1433,8 @@ usb3_phy1: usb-phy@382f0040 {
 
 		usb3_1: usb@32f10108 {
 			compatible = "fsl,imx8mp-dwc3";
-			reg = <0x32f10108 0x8>,
-			      <0x382f0000 0x20>;
+			reg = <0x0 0x32f10108 0x0 0x8>,
+			      <0x0 0x382f0000 0x0 0x20>;
 			clocks = <&clk IMX8MP_CLK_HSIO_ROOT>,
 				 <&clk IMX8MP_CLK_USB_ROOT>;
 			clock-names = "hsio", "suspend";
@@ -1445,7 +1448,7 @@ usb3_1: usb@32f10108 {
 
 			usb_dwc3_1: usb@38200000 {
 				compatible = "snps,dwc3";
-				reg = <0x38200000 0x10000>;
+				reg = <0x0 0x38200000 0x0 0x10000>;
 				clocks = <&clk IMX8MP_CLK_HSIO_AXI>,
 					 <&clk IMX8MP_CLK_USB_CORE_REF>,
 					 <&clk IMX8MP_CLK_USB_ROOT>;
@@ -1459,7 +1462,7 @@ usb_dwc3_1: usb@38200000 {
 
 		dsp: dsp@3b6e8000 {
 			compatible = "fsl,imx8mp-dsp";
-			reg = <0x3b6e8000 0x88000>;
+			reg = <0x0 0x3b6e8000 0x0 0x88000>;
 			mbox-names = "txdb0", "txdb1",
 				"rxdb0", "rxdb1";
 			mboxes = <&mu2 2 0>, <&mu2 2 1>,
